torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -216,16 +228,27 @@ def pre_forward(
     def post_forward(self, module: nn.Module, input: Any, output: Any):
         with torch.profiler.record_function("FSDP::post_forward"):
             self.reshard()
+            self._record_post_forward()
             self._training_state = TrainingState.IDLE
             return output
 
-    def pre_backward(self, *unused: Any):
+    def _record_post_forward(self) -> None:
+        # Since a group has one pre-backward unshard for each forward call
+        # before the backward, we record each usage (with multiplicity)
+        post_forward_index = len(self.comm_ctx.post_forward_order)
+        self.comm_ctx.post_forward_order.append(self)
+        self._post_forward_indices.append(post_forward_index)
+
+    def pre_backward(self, forward_grad_fns: Tuple[Any, ...], *unused: Any):
         with torch.profiler.record_function("FSDP::pre_backward"):
             self._training_state = TrainingState.PRE_BACKWARD
             self.unshard()  # no-op if prefetched
             self.wait_for_unshard()
+            # Can be already removed if running backward >1 time
```

Comment:
"running backward >1 time": is this still referring to the current loop (not accumulating grad) ?

Reply:
@awgu

Reply:
Sorry. This is referring to calling `.backward()` multiple times, where all but the last time would need to call with `retain_graph=True`.

Within a single `.backward()` call, I do not think we can run this multiple times.